### PR TITLE
API-1255 : Get back options labels in the product call

### PR DESCRIPTION
**Pull request by tomglvng**
Adds query parameter (no BC break and perf altered only if wanted) called “withAttributeOptions”Expected response is:

```
{
  "values": {
    "main_color": [
      {
        "data": "black",
        "linked_data": [
          {
            "code": "black"
          },
          {
            "attribute": "main_color"
          },
          {
            "labels": {
              "en_US": "Black",
              "fr_FR": "Noir"
            }
          }
        ]
      }
    ],
    "collection": [
      {
        "locale": null,
        "scope": null,
        "data": [
          "winter_2016",
          "summer_2016"
        ],
        "linked_data": {
          "winter_2016": {
            "attribute": "collection",
            "code": "winter_2016",
            "labels": {
              "en_US": "Winter 2016",
              "fr_FR": "Hiver 2016"
            }
          },
          "summer_2016": {
            "attribute": "collection",
            "code": "summer_2016",
            "labels": {
              "en_US": "Summer 2016",
              "fr_FR": "Eté 2016"
            }
          }
        }
      }
    ]
  }
}
```

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php [hunk 1]
```diff
@@ -143,6 +143,9 @@ class ProductController
     /** @var GetConnectorProducts */
     private $getConnectorProducts;
 
+    /** @var GetConnectorProducts */
+    private $getConnectorProductsWithLabels;
```
Comment (pierrejolly) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php:147]: `$getConnectorProductsWithOptionLabels` is better IMO

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php [hunk 2]
```diff
@@ -96,7 +103,11 @@ public function handle(ListProductsQuery $query): ConnectorProductList
 
         $pqb->addSorter('id', Directions::ASCENDING);
 
-        $connectorProductList = $this->getConnectorProductsQuery->fromProductQueryBuilder(
+        $connectorProductsQuery = 'true' === $query->withAttributeOptions ?
```
Comment (pierrejolly) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php:106]: You should use `$query-> withAttributeOptionsAsBoolean()`

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php (hunk 1 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php]: Done in 047289f2669f707cafd363913f8dd48e09a5fcd1

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php (hunk 2 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php]: Done in 047289f2669f707cafd363913f8dd48e09a5fcd1

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Query\ProductQueryBuilderInterface;
+use Akeneo\Pim\Enrichment\Component\Product\Value\OptionsValue;
+use Akeneo\Pim\Enrichment\Component\Product\Value\OptionValue;
+use Doctrine\DBAL\Connection;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SqlGetConnectorProductsWithOptionLabels implements Query\GetConnectorProducts
```
Comment (pchasle) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php:19]: Missing integration test

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        }
+
+        return $optionCodes;
+    }
+
+    /**
+     * @param array $optionCodes [['attribute_code', 'option_code']]
+     *
+     * @return array ['attribute_code' => ['option_code' => ['en_US' => 'translation']]
```
Comment (pchasle) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php:93]: You can use the phpstan type notation instead, that will allow phpstan to do some static analysis on those kind of phpdoc
See https://phpstan.org/writing-php-code/phpdoc-types#array-shapes

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php [hunk 5]
```diff
@@ -45,6 +45,9 @@ final class ListProductsQueryHandler
     /** @var GetConnectorProducts */
     private $getConnectorProductsQuery;
 
+    /** @var GetConnectorProducts */
+    private $getConnectorProductsQuerywithOptionLabel;
```
Comment (fitn) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php:49]:
```suggestion
    private $getConnectorProductsQueryWithOptions;
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php [hunk 6]
```diff
@@ -58,6 +61,7 @@ public function __construct(
         ProductQueryBuilderFactoryInterface $searchAfterPqbFactory,
         PrimaryKeyEncrypter $primaryKeyEncrypter,
         GetConnectorProducts $getConnectorProductsQuery,
+        GetConnectorProducts $getConnectorProductsQuerywithOptionLabel,
```
Comment (fitn) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php:64]:
```suggestion
        GetConnectorProducts $getConnectorProductsQueryWithOptions,
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php [hunk 7]
```diff
@@ -66,11 +70,13 @@ public function __construct(
         $this->searchAfterPqbFactory = $searchAfterPqbFactory;
         $this->primaryKeyEncrypter = $primaryKeyEncrypter;
         $this->getConnectorProductsQuery = $getConnectorProductsQuery;
+        $this->getConnectorProductsQuerywithOptionLabel = $getConnectorProductsQuerywithOptionLabel;
```
Comment (fitn) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php:73]:
```suggestion
        $this->getConnectorProductsQueryWithOptions = $getConnectorProductsQueryWithOptions;
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php [hunk 8]
```diff
@@ -96,7 +102,11 @@ public function handle(ListProductsQuery $query): ConnectorProductList
 
         $pqb->addSorter('id', Directions::ASCENDING);
 
-        $connectorProductList = $this->getConnectorProductsQuery->fromProductQueryBuilder(
+        $connectorProductsQuery = $query->withAttributeOptionsAsBoolean() ?
+            $this->getConnectorProductsQuerywithOptionLabel :
```
Comment (fitn) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php:106]:
```suggestion
            $this->getConnectorProductsQueryWithOptions :
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+namespace Akeneo\Pim\Enrichment\Component\Product\Value;
+
+use Akeneo\Pim\Enrichment\Component\Product\Model\AbstractValue;
+use Akeneo\Pim\Enrichment\Component\Product\Model\ValueInterface;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class OptionValueWithLabels extends AbstractValue implements OptionValueInterface
```
[the remark this answers is not in the document]
Comment (fitn) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php:12]: And not the right class name

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+namespace Akeneo\Pim\Enrichment\Component\Product\Value;
+
+use Akeneo\Pim\Enrichment\Component\Product\Model\AbstractValue;
+use Akeneo\Pim\Enrichment\Component\Product\Model\ValueInterface;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class OptionsValueWithLabels extends AbstractValue
```
[the remark this answers is not in the document]
Comment (fitn) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php:12]: not the right class name

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/UseCase/ListProductsQueryHandlerSpec.php [hunk 11]
```diff
@@ -40,6 +41,7 @@ function let(
             $searchAfterPqbFactory,
             $primaryKeyEncrypter,
             $getConnectorProducts,
+            $getConnectorProductswithOptionLabel,
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/UseCase/ListProductsQueryHandlerSpec.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public function isEqual(ValueInterface $value): bool
+    {
+        if (!$value instanceof OptionValueWithLabels ||
+            $this->getScopeCode() !== $value->getScopeCode() ||
+            $this->getLocaleCode() !== $value->getLocaleCode()) {
+            return false;
+        }
+
+        // useless for POC but inaccurate
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php]: To add: thanks to spec tests ;)

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public function isEqual(ValueInterface $value): bool
+    {
+        if (!$value instanceof OptionValueWithLabels ||
+            $this->getScopeCode() !== $value->getScopeCode() ||
+            $this->getLocaleCode() !== $value->getLocaleCode()) {
+            return false;
+        }
+
+        // useless for POC but inaccurate
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php]: Same correction

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionValueWithLabels.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Value/OptionsValueWithLabels.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/ListProducts/SuccessListProductEndToEnd.php [hunk 14]
```diff
@@ -972,7 +1002,54 @@ private function getStandardizedProducts(): array
...
+                            "data": "optionA",
+                            "linked_data": {
+                                "attribute": "a_simple_select",
+                                "code": "optionA",
+                                "labels": {
+                                    "en_US": "Option A"
+                                }
+                            }
+                        }
+                    ],
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/ListProducts/SuccessListProductEndToEnd.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php (hunk 4 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php [hunk 15]
```diff
@@ -143,6 +143,9 @@ class ProductController
     /** @var GetConnectorProducts */
     private $getConnectorProducts;
 
+    /** @var GetConnectorProducts */
+    private $getConnectorProductsWithOptionLabels;
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php [hunk 16]
```diff
@@ -184,6 +187,7 @@ public function __construct(
         ConnectorProductNormalizer $connectorProductNormalizer,
         TokenStorageInterface $tokenStorage,
         GetConnectorProducts $getConnectorProducts,
+        GetConnectorProducts $getConnectorProductsWithOptionLabels,
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php [hunk 17]
```diff
@@ -215,6 +219,7 @@ public function __construct(
         $this->connectorProductNormalizer = $connectorProductNormalizer;
         $this->tokenStorage = $tokenStorage;
         $this->getConnectorProducts = $getConnectorProducts;
+        $this->getConnectorProductsWithOptionLabels = $getConnectorProductsWithOptionLabels;
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php [hunk 18]
```diff
@@ -286,15 +292,21 @@ public function listAction(Request $request): JsonResponse
...
-     * @throws NotFoundHttpException
+     * @param Request $request
+     * @param string $code
+     * @return JsonResponse
      */
-    public function getAction(string $code): JsonResponse
+    public function getAction(Request $request, string $code): JsonResponse
     {
+        $connectorProductsQuery = 'true' === $request->query->get('with_attribute_options', "false") ?
+            $this->getConnectorProductsWithOptionLabels :
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/ProductController.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/config/connector/use_cases.yml [hunk 19]
```diff
@@ -25,6 +25,7 @@ services:
             - '@pim_catalog.query.product_query_builder_search_after_size_factory_external_api'
             - '@pim_api.security.primary_key_encrypter'
             - '@akeneo.pim.enrichment.product.connector.get_product_from_identifiers'
+            - '@akeneo.pim.enrichment.product.connector.get_product_from_identifiers_with_option_labels'
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Resources/config/connector/use_cases.yml]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php (hunk 6 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php (hunk 7 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php (hunk 8 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/UseCase/ListProductsQueryHandler.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php (hunk 3 above)
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php]: Done in 220994d384f63bd5896657cb746f93c15a33cff6

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+        return $this->getConnectorProductsWithLabels([$connectorProduct])[0];
+    }
+
+    private function getConnectorProductsWithLabels(array $connectorProducts): array
+    {
+        $optionCodes = $this->getOptionCodes($connectorProducts);
+        $optionWithLabels = $this->getOptionWithLabels($optionCodes);
+
+        return array_map(function (ConnectorProduct $product) use ($optionWithLabels) {
+            return $product->transformOptionWithLabels($optionWithLabels);
```
[the remark this answers is not in the document]
Comment (tomglvng) [reply src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/SqlGetConnectorProductsWithOptionLabels.php]: True. I renamed it to `buildLinkedData`

**Review by willy-ahva — APPROVED** [review]

**Review by tomglvng — COMMENTED** [review]
master is frozen

**Pull request merged**